### PR TITLE
Adding a few more tests about WINDOW_UPDATE edge cases

### PR DESCRIPTION
**Pull request by jrudolph**

### CODE REVIEW

File: akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ServerSpec.scala [hunk 1]
```diff
@@ -452,6 +453,66 @@ class Http2ServerSpec extends AkkaSpecWithMaterializer("""
...
+        sendWINDOW_UPDATE(0, missingWindow) // SETTINGS_INITIAL_WINDOW_SIZE only has effect on stream-level window, so need to make room in connection
+
+        entityDataOut.sendNext(bytes(bytesToSend, 0x23))
+        expectDATA(TheStreamId, false, Http2Protocol.InitialWindowSize)
+        expectNoBytes(100.millis)
+
+        sendSETTING(Http2Protocol.SettingIdentifier.SETTINGS_INITIAL_WINDOW_SIZE, bytesToSend)
+        updateFromServerWindows(TheStreamId, _ + missingWindow) // test probe doesn't automatically update window
+        expectDATA(TheStreamId, false, missingWindow)
+        expectSettingsAck() // FIXME: bug: we must send ACK before making use of the new setting, see https://github.com/akka/akka-http/issues/3553
```
Comment (jrudolph) [inline akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ServerSpec.scala:510]: @ignasi35 this test found the issue around applying settings we discussed in #3553.

File: akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ServerSpec.scala [hunk 2]
```diff
@@ -452,6 +453,66 @@ class Http2ServerSpec extends AkkaSpecWithMaterializer("""
         toNet.expectNoBytes(100.millis) // the whole stage failed with bug #2236
       }
 
+      "handle WINDOW_UPDATE correctly when received before started sending out response and while receiving request data" in new WaitingForRequestData {
+        val bytesToSend = 70000 // > Http2Protocol.InitialWindowSize
+        val missingWindow = bytesToSend - Http2Protocol.InitialWindowSize
+        require(missingWindow >= 0)
+        sendWINDOW_UPDATE(0, missingWindow)
```
Comment (raboof) [inline akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ServerSpec.scala:460]: So this extends the window size with `missingWindow` bytes beyond the `InitialWindowSize`. Makes sense, but perhaps a comment might make this easier to follow ;)

File: akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ServerSpec.scala [hunk 3]
```diff
@@ -452,6 +453,66 @@ class Http2ServerSpec extends AkkaSpecWithMaterializer("""
         toNet.expectNoBytes(100.millis) // the whole stage failed with bug #2236
       }
 
+      "handle WINDOW_UPDATE correctly when received before started sending out response and while receiving request data" in new WaitingForRequestData {
```
Comment (raboof) [inline akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ServerSpec.scala:456]: Does the request never finish? OK I guess :)

File: akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ServerSpec.scala [hunk 4]
```diff
@@ -452,6 +453,66 @@ class Http2ServerSpec extends AkkaSpecWithMaterializer("""
...
+        expectDecodedResponseHEADERS(streamId = TheStreamId, endStream = false) shouldBe response.withEntity(HttpEntity.Empty.withContentType(ContentTypes.`application/octet-stream`))
+        entityDataOut.sendNext(bytes(bytesToSend, 0x23))
+
+        expectDATA(TheStreamId, false, bytesToSend)
+
+        entityDataOut.sendComplete()
+        expectDATA(TheStreamId, true, 0)
+      }
+
+      "handle WINDOW_UPDATE correctly when received before started sending out response" in new WaitingForResponseSetup {
```
Comment (raboof) [inline akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ServerSpec.scala:476]: Very similar to the previous one but now with the request already finished, seems OK

File: akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ServerSpec.scala (hunk 1 above)
Comment (raboof) [reply akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ServerSpec.scala:510]: So when this bug is fixed we should see the Ack before we see the DATA? Make sense.

**Review by raboof — APPROVED** [review]

File: akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ServerSpec.scala (hunk 2 above)
Comment (ignasi35) [reply akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ServerSpec.scala:460]: > perhaps a comment might make this easier to follow

I came here just to write that :-)

File: akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ServerSpec.scala (hunk 1 above)
Comment (ignasi35) [reply akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ServerSpec.scala:510]: I'm not sure I'm familiar enought with the issue. In cases when the window is exhausted, no frames should be produced and buffered until (1) a `WINDOW_UPDATE` is received, or (2) a `SETTINGS` is received and the corresponding `SETTING_ACK` is sent out.

When #3553 is fixed, I agree that `SETTINGS_ACK` should arrive before the SETTINGS ACK.

The problem is, when the server window on the client side is exhausted, the client should not pre-render any frames until a WINDOW_UPDATE (or a SETTINGS_INITIAL_WINDOW_SIZE) is received since there are no guarantees a pre-rendered frame would fit on the window, right? The same happens with SETTINGS_FRAME_SIZE.

------

Hmm, increasing the window size via settings is very different from the processing of a `WINDOW_UPDATE`. 

In `WINDOW_UPDATE`, the sender increases the window value and then sends the frame out so the remote peer's value for the window is always less or equal to the sender's value.
Using `SETTINGS_INITIAL_WINDOW_SIZE` means the sender doesn't increase the window value until it gets an `SETTINGS_ACK` but the remote peer already uses the increased window value so the remote peer's value for the window may be greater than the sender's window value.

It sounds like `SETTINGS_INITIAL_WINDOW_SIZE` as a broadcast replacement of `WINDOW_UPDATE` is a hack.

File: akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ServerSpec.scala [hunk 5]
```diff
@@ -452,6 +453,66 @@ class Http2ServerSpec extends AkkaSpecWithMaterializer("""
...
+        require(missingWindow >= 0)
+        sendWINDOW_UPDATE(0, missingWindow)
+        sendWINDOW_UPDATE(TheStreamId, missingWindow)
+
+        val entityDataOut = TestPublisher.probe[ByteString]()
+
+        val response = HttpResponse(entity = HttpEntity(ContentTypes.`application/octet-stream`, Source.fromPublisher(entityDataOut)))
+        emitResponse(TheStreamId, response)
+        expectDecodedResponseHEADERS(streamId = TheStreamId, endStream = false) shouldBe response.withEntity(HttpEntity.Empty.withContentType(ContentTypes.`application/octet-stream`))
+        entityDataOut.sendNext(bytes(bytesToSend, 0x23))
```
Comment (ignasi35) [inline akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ServerSpec.scala:468]: How does this work? 
First, the server emits a response, and then it sends `bytesToSend` bytes? Doesn't that exceed the window size of `bytesToSend` octets?

File: akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ServerSpec.scala (hunk 4 above)
Comment (ignasi35) [reply akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ServerSpec.scala:476]: I need to read more about `WaitingForResponseSetup` and `WaitingForRequestData` to see the difference between these two tests.

File: akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ServerSpec.scala (hunk 1 above)
Comment (jrudolph) [reply akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ServerSpec.scala:510]: > The problem is, when the server window on the client side is exhausted, the client should not pre-render any frames until a WINDOW_UPDATE (or a SETTINGS_INITIAL_WINDOW_SIZE) is received since there are no guarantees a pre-rendered frame would fit on the window, right? The same happens with SETTINGS_FRAME_SIZE.

Yes.

> In `WINDOW_UPDATE`, the sender increases the window value and then sends the frame out so the remote peer's value for the window is always less or equal to the sender's value.
> Using `SETTINGS_INITIAL_WINDOW_SIZE` means the sender doesn't increase the window value until it gets an `SETTINGS_ACK` but the remote peer already uses the increased window value so the remote peer's value for the window may be greater than the sender's window value.

For the right value of sender (i.e. the one that proposed the setting), that seems correct as well.

> It sounds like `SETTINGS_INITIAL_WINDOW_SIZE` as a broadcast replacement of `WINDOW_UPDATE` is a hack.

Agreed, that's not what it is supposed to be used for. I guess it was really originally thought to be used for setting the initial window size for new streams. Maybe it was changed this way because the sender can just ignore the SETTINGS ACK and assume the setting to be applied instantly and retroactively, so that it doesn't matter for the client if a received stream was created slightly before the corresponding ACK or slightly after.

File: akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ServerSpec.scala [hunk 6]
```diff
@@ -452,6 +453,66 @@ class Http2ServerSpec extends AkkaSpecWithMaterializer("""
...
+        val missingWindow = bytesToSend - Http2Protocol.InitialWindowSize
+        require(missingWindow >= 0)
+        sendWINDOW_UPDATE(0, missingWindow) // SETTINGS_INITIAL_WINDOW_SIZE only has effect on stream-level window, so need to make room in connection
+
+        entityDataOut.sendNext(bytes(bytesToSend, 0x23))
+        expectDATA(TheStreamId, false, Http2Protocol.InitialWindowSize)
+        expectNoBytes(100.millis)
+
+        sendSETTING(Http2Protocol.SettingIdentifier.SETTINGS_INITIAL_WINDOW_SIZE, bytesToSend)
+        updateFromServerWindows(TheStreamId, _ + missingWindow) // test probe doesn't automatically update window
```
[the remark this answers is not in the document]
Comment (jrudolph) [reply akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ServerSpec.scala:508]: No, I think that doesn't work because we consumed the initial part of the window already.

File: akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ServerSpec.scala (hunk 5 above)
Comment (jrudolph) [reply akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ServerSpec.scala:468]: The user level interface uses a simple `Source[ByteString]` which can send any amount of data (I talked about that in our call this morning). The internal logic for outgoing streams is this:

 * frames are pulled one-by-one by the lower layers from the Http2ServerDemux
 * we keep a buffer of control frames to send (control frames are not subject to flow-control / windows) and a list of streams that can send data
 * if the multiplexer is pulled, first the control frames are drained and only if there are no pending control frames, one ready stream is chosen to send the next data frame
 * for each stream we have an outgoing buffer and if there's pending data the stream is listed to be chosen to send a frame
 * into that buffer, we pull from the user until we have at least `2 * currentMaxFrameSize` number of bytes buffered

File: akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ServerSpec.scala (hunk 4 above)
Comment (jrudolph) [reply akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ServerSpec.scala:476]: Yep, two different states in the stream state machine..

File: akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ServerSpec.scala (hunk 2 above)
Comment (jrudolph) [reply akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ServerSpec.scala:461]: Done

File: akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ServerSpec.scala (hunk 5 above)
Comment (jrudolph) [reply akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ServerSpec.scala:469]: So, in case that wasn't clear, here we use the user-level interface to send a ByteString bigger than the WINDOW_SIZE and expect the rest to be buffered and only flushed later when more WINDOW is available.

Comment (raboof) [reply akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ServerSpec.scala:469]: `entityDataOut` is a probe stream that is attached to the body of the `HttpResponse`. Only DATA frames count to the window size (https://tools.ietf.org/html/rfc7540#section-5.2.1 point 5), so this fits exactly
